workbench-core/environments/src/handlers/environmentTypeHandler.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ProductViewSummary, ProvisioningArtifactDetail } from '@aws-sdk/client-service-catalog';
+import { AwsService, resourceTypeToKey, CFNTemplateParameters } from '@aws/workbench-core-base';
+import { EnvironmentType } from '../models/environmentTypes/environmentType';
+import EnvironmentTypeService from '../services/environmentTypeService';
+
+export default class EnvironmentTypeHandler {
+  private _mainAccountAwsService: AwsService;
+
+  public constructor(mainAccountAwsService: AwsService) {
+    this._mainAccountAwsService = mainAccountAwsService;
+  }
+  /* eslint-disable-next-line @typescript-eslint/no-explicit-any*/
+  public async execute(event: any): Promise<void> {
+    console.log(`Processing Environment Types`);
+    const portfolioName = process.env.SC_PORTFOLIO_NAME!;
+    const portfolioId = await this._mainAccountAwsService.helpers.serviceCatalog.getPortfolioId(
+      portfolioName
+    );
+    if (portfolioId === undefined) {
+      throw new Error(`Could not find portfolioId for portfolio: ${portfolioName}`);
+    }
+    console.log(`Processing portfolio ${portfolioId}`);
+    const products = await this._mainAccountAwsService.helpers.serviceCatalog.getProductsByPortfolioId(
+      portfolioId
+    );
+    for (const product of products) {
+      try {
+        console.log(`Processing product ${product.ProductId}`);
+        const currentProductId = product.ProductId || '';
+        const provisionArtifacts =
+          await this._mainAccountAwsService.helpers.serviceCatalog.getProvisionArtifactsByProductId(
+            currentProductId
+          );
+        for (const provisionArtifact of provisionArtifacts.filter((p) => p.Active)) {
```

Comment:
Is O(N^2) complexity our best solution for this? Is there any way to bring it down to linear time?

Reply:
unfortunately I didn't find better options from the api , the best solution for this would be creating a listener for provision artifacts and create every time something new is added to the catalog and would be something great to explore after GA